### PR TITLE
Upgradable warp routes

### PR DESCRIPTION
**Pull request by ltyu**
### Description

Makes the following Warp Routes upgradable through CLI deployment
```
- FastHypERC20Collateral
- FastHypERC20
- HypERC20
- HypERC20Collateral
- HypERC20CollateralVaultDeposit
- HypNative
- HypNativeScaled
- HypERC721Collateral
- HypERC721
- HypERC721URICollateral
- HypERC721URIStorage
```

- Adds `initialize()` to each contract
	- It mostly contains `_MailboxClient_initialize`, and any additional proxy specific constructor logic
- Refactors
	- Update `GasRouterDeployer` to inherit from `ProxiedRouterDeployer`
	- Update `ProxiedRouterDeployer.routerContractName` to  `abstract ProxiedRouterDeployer.routerContractName(): RouterKey` .
		- This allows child classes to specify their own contract name instead of being locked-in to a single name upon construction. Similar concept to the existing `ProxiedRouterDeployer.constructorArgs()` and `ProxiedRouterDeployer.initializeArgs()`
	- Update `router()` 
		- Add function into `InterchainQueryDeployer`
		- Add function into `InterchainAccountDeployer`
		- Update to abstract function in `ProxiedRouterDeployer`
		- ### Drive-by changes

### Related issues
Fixes #999 

### Backward compatibility
- Should be backward compatible

### Testing
- Updates contract unit tests
- Manual testing through CLI and verified contracts

### CODE REVIEW

**Review by nbayindirli — COMMENTED** [review]
Curious why we're using Transparent over UUPS? UUPS tends to be the de-facto recommendation these days (cheaper gas, more flexible, can remove eventually remove upgradability)

File: typescript/sdk/src/middleware/account/InterchainAccountDeployer.ts [hunk 1]
```diff
@@ -30,14 +30,26 @@ export class InterchainAccountDeployer extends ProxiedRouterDeployer<
     });
   }
 
-  async constructorArgs(_: string, config: RouterConfig): Promise<[string]> {
+  routerContractName<K extends keyof InterchainAccountFactories>(
```
Comment (yorhodes) [inline typescript/sdk/src/middleware/account/InterchainAccountDeployer.ts:33]: why have this take in config if its unused?

File: typescript/sdk/src/token/deploy.ts [hunk 2]
```diff
@@ -65,6 +59,51 @@ export class HypERC20Deployer extends GasRouterDeployer<
...
+        isFastConfig(config) ? TokenType.fastCollateral : TokenType.collateral
+      ) as K;
+    } else if (isSyntheticConfig(config)) {
+      return (
+        isFastConfig(config) ? TokenType.fastSynthetic : TokenType.synthetic
+      ) as K;
+    } else {
+      throw new Error('Unknown collateral type when constructing router name');
+    }
+  }
```
Comment (yorhodes) [inline typescript/sdk/src/token/deploy.ts:76]: oh I see, interesting

Comment (yorhodes) [reply typescript/sdk/src/token/deploy.ts:76]: I think @paulbalaji also changed this in https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/3490

File: typescript/sdk/src/token/deploy.ts [hunk 3]
```diff
@@ -65,6 +59,51 @@ export class HypERC20Deployer extends GasRouterDeployer<
...
+    if (isCollateralConfig(config)) {
+      return [config.token, config.mailbox];
+    } else if (isNativeConfig(config)) {
+      return config.scale ? [config.scale, config.mailbox] : [config.mailbox];
+    } else if (isSyntheticConfig(config)) {
+      return [config.decimals, config.mailbox];
+    } else {
+      throw new Error('Unknown collateral type when constructing arguments');
+    }
+  }
```
Comment (yorhodes) [inline typescript/sdk/src/token/deploy.ts:88]: I think you can strongly type this if you infer the `factory` type from the `config.tokenType`

File: typescript/sdk/src/token/deploy.ts [hunk 4]
```diff
@@ -65,6 +59,51 @@ export class HypERC20Deployer extends GasRouterDeployer<
...
+
+  async initializeArgs(_: ChainName, config: ERC20RouterConfig): Promise<any> {
```
Comment (yorhodes) [inline typescript/sdk/src/token/deploy.ts:90]: ditto for strongly typing this rather than `any`

File: typescript/sdk/src/token/deploy.ts [hunk 5]
```diff
@@ -131,91 +170,20 @@ export class HypERC20Deployer extends GasRouterDeployer<
...
-
   router(contracts: HyperlaneContracts<HypERC20Factories>) {
     for (const key of objKeys(hypERC20factories)) {
       if (contracts[key]) {
-        return contracts[key];
+        return contracts[key] as GasRouter;
       }
     }
     throw new Error('No matching contract found');
   }
```
Comment (yorhodes) [inline typescript/sdk/src/token/deploy.ts:180]: maybe this should take the config too? and then we can do `contracts[config.tokenType]`

File: typescript/sdk/src/token/deploy.ts [hunk 6]
```diff
@@ -131,91 +170,20 @@ export class HypERC20Deployer extends GasRouterDeployer<
...
-    } else if (isSyntheticConfig(config)) {
-      router = await this.deploySynthetic(chain, config);
-    } else {
-      throw new Error('Invalid ERC20 token router config');
-    }
-    await this.configureClient(chain, router, config);
+    const { [this.routerContractName(config)]: router } =
+      await super.deployContracts(chain, config);
+
+    await this.configureClient(chain, router as MailboxClient, config);
```
Comment (yorhodes) [inline typescript/sdk/src/token/deploy.ts:186]: im surprised you have to cast here?

File: typescript/sdk/src/token/deploy.ts [hunk 7]
```diff
@@ -292,6 +260,48 @@ export class HypERC721Deployer extends GasRouterDeployer<
       contractVerifier,
     });
   }
+  routerContractName<K extends keyof HypERC721Factories>(
+    config: ERC721RouterConfig,
+  ): K {
+    if (isCollateralConfig(config)) {
+      return (
+        isUriConfig(config) ? TokenType.collateralUri : TokenType.collateral
+      ) as K;
```
Comment (yorhodes) [inline typescript/sdk/src/token/deploy.ts:269]: do you have to cast here?

**Review by yorhodes — APPROVED** [review]
nice work
glad we reused `ProxiedRouterDeployer`

File: typescript/sdk/src/token/deploy.ts (hunk 2 above)
Comment (ltyu) [reply typescript/sdk/src/token/deploy.ts:76]: I think Paul's PR will compliment this. I've noted to merge his first so i can rebase

File: typescript/sdk/src/middleware/account/InterchainAccountDeployer.ts (hunk 1 above)
Comment (ltyu) [reply typescript/sdk/src/middleware/account/InterchainAccountDeployer.ts:33]: It's overriding the abstract so keeping the same sig

File: solidity/contracts/token/HypERC721.sol [hunk 8]
```diff
@@ -23,7 +23,10 @@ contract HypERC721 is ERC721EnumerableUpgradeable, TokenRouter {
...
         uint256 _mintAmount,
         string memory _name,
-        string memory _symbol
+        string memory _symbol,
+        address _hook,
+        address _interchainSecurityModule,
+        address _owner
     ) external initializer {
         address owner = msg.sender;
         _transferOwnership(owner);
```
[the remark this answers is not in the document]
Comment (ltyu) [reply solidity/contracts/token/HypERC721.sol:32]: Fixed. Cleaned this a little bit by putting `_Mailboxclient` at the top, and calling `_safeMint(msg.sender)`

File: typescript/sdk/src/middleware/account/InterchainAccountDeployer.ts [hunk 9]
```diff
@@ -30,14 +30,26 @@ export class InterchainAccountDeployer extends ProxiedRouterDeployer<
...
+    return 'interchainAccountRouter' as K;
+  }
+
+  router(contracts: HyperlaneContracts<InterchainAccountFactories>): Router {
+    for (const key of objKeys(interchainAccountFactories)) {
+      if (contracts[key]) {
+        return contracts[key] as Router;
+      }
+    }
+    throw new Error('No matching contract found');
```
[the remark this answers is not in the document]
Comment (ltyu) [reply typescript/sdk/src/middleware/account/InterchainAccountDeployer.ts]: Fixed.

File: typescript/sdk/src/middleware/liquidity-layer/LiquidityLayerRouterDeployer.ts [hunk 10]
```diff
@@ -67,17 +71,29 @@ export class LiquidityLayerDeployer extends ProxiedRouterDeployer<
...
-    config: LiquidityLayerConfig,
-  ): Promise<[string]> {
+  routerContractName<K extends keyof LiquidityLayerFactories>(
+    _: RouterConfig,
+  ): K {
+    return 'liquidityLayerRouter' as K;
+  }
+
+  router(contracts: HyperlaneContracts<LiquidityLayerFactories>): Router {
+    for (const key of objKeys(liquidityLayerFactories)) {
```
[the remark this answers is not in the document]
Comment (ltyu) [reply typescript/sdk/src/middleware/liquidity-layer/LiquidityLayerRouterDeployer.ts]: Fixed.

File: typescript/sdk/src/token/deploy.ts (hunk 5 above)
Comment (ltyu) [reply typescript/sdk/src/token/deploy.ts:204]: Same as above. It's overriding the abstract so keeping the same sig.

**Pull request merged**